In [11]:
import pandas as pd
import numpy as np


In [46]:
df = pd.read_csv("train.csv")

In [47]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [48]:
df.dropna(inplace = True)df.drop

In [49]:
df

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

In [50]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
df.drop("Name", axis =1, inplace = True)
df.drop("PassengerId", axis = 1, inplace = True)
df.drop("Ticket", axis = 1,inplace = True)
df.drop("Fare", axis = 1, inplace = True)
X = df.drop("Survived", axis = 1)
y = df["Survived"]
categorical_features = ["Sex","Cabin", "Embarked"]
one_hot= OneHotEncoder()
transformer = ColumnTransformer([('one_hot',one_hot,categorical_features)], remainder = "passthrough")
transformed_X = transformer.fit_transform(X) 


In [51]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
# Keep the default hyperparameters

# Fit the model to the training data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(transformed_X,y, test_size=0.2)

In [52]:
clf.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [61]:
y_preds = clf.predict(X_test)

In [54]:
clf.score(X_train,y_train)


1.0

In [55]:
clf.score(X_test,y_test)
import pickle
pickle.dump(clf, open("random_forest_Titanic_2.pkl","wb"))


In [56]:
clf.score(X_test,y_test) 

0.5405405405405406

# Different results:
### random_forest_Titanic_1.pkl
score = 0.729

## After deleting Fare and Ticket columns
## random_forest_Titanic_1.pkl
score = 0.540


In [59]:
loaded_model = pickle.load(open("random_forest_Titanic_1.pkl","rb"))


In [60]:
loaded_model.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [64]:
y_preds = loaded_model.predict(X_test)

In [65]:
loaded_model.score(X_train,y_train)


1.0

In [66]:
loaded_model.score(X_test,y_test) 

0.5675675675675675

## Evaluating the model on dataset w/out Fare and Ticket Columns


In [67]:
import numpy as np
# Improve the model
# Try different ammounts of estimators
np.random.seed(40)
for i in range(10,100,10):
    print(f"Trying model with {i} estimators...")
    loaded_model = RandomForestClassifier(n_estimators = i).fit(X_train, y_train)
    print(f"Model accuracy on the test set: {clf.score(X_test,y_test) * 100:.2f}")
    print("")

Trying model with 10 estimators...
Model accuracy on the test set: 54.05

Trying model with 20 estimators...
Model accuracy on the test set: 54.05

Trying model with 30 estimators...
Model accuracy on the test set: 54.05

Trying model with 40 estimators...
Model accuracy on the test set: 54.05

Trying model with 50 estimators...
Model accuracy on the test set: 54.05

Trying model with 60 estimators...
Model accuracy on the test set: 54.05

Trying model with 70 estimators...
Model accuracy on the test set: 54.05

Trying model with 80 estimators...
Model accuracy on the test set: 54.05

Trying model with 90 estimators...
Model accuracy on the test set: 54.05



### The score is unchanging, score = 54.05
##### Next training and evaluation of the model is done on the initial dataset - that is with Fare and Ticket columns 

In [79]:
df = pd.read_csv("train.csv")


In [80]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [81]:
df.dropna(inplace=True) # droping rows with missing data

In [82]:
df

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

In [83]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
df.drop("Name", axis =1, inplace = True)
df.drop("PassengerId", axis = 1, inplace = True)

X = df.drop("Survived", axis = 1)
y = df["Survived"]
categorical_features = ["Sex","Cabin","Ticket","Embarked"]
one_hot= OneHotEncoder()
transformer = ColumnTransformer([('one_hot',one_hot,categorical_features)], remainder = "passthrough")
transformed_X = transformer.fit_transform(X) 

In [89]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(transformed_X,y, test_size=0.2)
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()

In [90]:
clf.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [91]:
clf.score(X_train,y_train)

1.0

In [92]:
clf.score(X_test,y_test) 

0.7567567567567568

In [93]:
import numpy as np
# Improve the loaded and retrained model to match the previous dataset
# Trying different ammounts of estimators
np.random.seed(40)
for i in range(10,100,10):
    print(f"Trying model with {i} estimators...")
    clf = RandomForestClassifier(n_estimators = i).fit(X_train, y_train)
    print(f"Model accuracy on the test set: {clf.score(X_test,y_test) * 100:.2f}")
    print("")

Trying model with 10 estimators...
Model accuracy on the test set: 64.86

Trying model with 20 estimators...
Model accuracy on the test set: 72.97

Trying model with 30 estimators...
Model accuracy on the test set: 70.27

Trying model with 40 estimators...
Model accuracy on the test set: 67.57

Trying model with 50 estimators...
Model accuracy on the test set: 75.68

Trying model with 60 estimators...
Model accuracy on the test set: 70.27

Trying model with 70 estimators...
Model accuracy on the test set: 70.27

Trying model with 80 estimators...
Model accuracy on the test set: 72.97

Trying model with 90 estimators...
Model accuracy on the test set: 75.68



#### The best score after evalution is with 90 estimators and is matching the current model score - that is 75.68
